Added validation of css color value

dcgit opened the pull request:
<!-- FreeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/FreeCodeCamp/FreeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->

<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->

<!-- Make sure that your PR is not a duplicate -->
#### Pre-Submission Checklist

<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->

<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->

<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #11226
#### Description

<!-- Describe your changes in detail -->

I updated the regex to include validation for allowable values that make the text white including the keyword white as well as hex #ffffff. Additional consideration for typos and extra space/characters that may cause the browser to ignore the color keyword. e.g. if the user types colors:white it will fail validation. I understand there is an existing pull request for this issue but I believe this is more robust.

Commit:
Added validation of css color value

Bouncey commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -4584,7 +4584,7 @@
         "assert($(\"h1\").hasClass(\"blue-text\"), 'message: Your <code>h1</code> element should have the class <code>blue-text</code>.');",
         "assert($(\"h1\").attr(\"id\") === \"orange-text\", 'message: Your <code>h1</code> element should have the id of <code>orange-text</code>.');",
-        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style[\\s\\S]*?color\\s*?:/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
+        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style\\s*?=\\s*(\"|'){1}\\s*(color|([\\s\\S]*;{1})*\\s*color)\\s*?:\\s*?(white|#ffffff)\\s*(;|'|\")/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
```
The test asks for `color: white`. Please remove the hex code

We don't need capturing groups here, please remove them

dcgit committed:
Removed hex and grouping

It was requested that hex not be used in validation so it was removed. Only the keyword white will be accepted.

dcgit committed:
fix/validate color is set to white

Removed hex and groupings

dcgit committed:
fix/css color white check

Removed hex and groupings

dcgit commented:
Made requested corrections.

systimotic commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -4584,7 +4584,7 @@
         "assert($(\"h1\").hasClass(\"blue-text\"), 'message: Your <code>h1</code> element should have the class <code>blue-text</code>.');",
         "assert($(\"h1\").attr(\"id\") === \"orange-text\", 'message: Your <code>h1</code> element should have the id of <code>orange-text</code>.');",
-        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style[\\s\\S]*?color\\s*?:/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
+        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style\\s*?=\\s*[\\\"]{1}\\s*\\s*\\S*;*\\s*color\\s*?:\\s*?white\\s*;*\\s*[\\\"]{1}/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
```
There are still some issues with this PR.
The majority of the issues have to do with the matching of spaces. `\s*` will match zero or more whitespace characters. Anywhere where whitespace is allowed but not required, please use that. 
Currently, a space between `h1` and `style` is optional, meaning that `<h1style=` will pass. `+` matches 1 or more whitespaces, so `\s+` should do the trick there.
I feel confident using RegEx, but I'm not a mastermind just yet. Can you please share what you're trying to do with `[\\\"]{1}`? From what I understand, you should only need to match a quote there.
With this change, I think we no longer need both of the `code.match`'s in the test. You can remove the first one.
Thank you for contributing!

Bouncey replied:
If I use single quotes like `'color: white'`, the test will fail.

We do not need the curly-braced quantifies for this.

 `\s{1}` is the exact same as `\s`.

The curly-braced quantifiers are handy if you need to match an exact number of something, no more, no less. `\s{10}` is more readable than `\s\s\s\s\s\s\s\s\s\s`

Bouncey replied:
@systimotic these tests are parsed from strings, so `\\\"` allows for an escaped back-slash, and an escaped double-quote. The quantifier is not needed, see my review :point_down:

systimotic replied:
Sorry for the messy email reply, not sure what happened there. I got as far as realizing that it would match either a backslash or a double quote in the end. What I realized after your comment is that because it ends up as `\"`, it doesn't match a backslash and a quote, but it's an escaped quote, so it just matches a quote. However, as far as I'm aware, quotes don't have to be escaped in a RegEx. I also don't understand why it's in a character set (the `[]`). http://codepen.io/Timoh/pen/vXqvZV?editors=0012

systimotic commented:
Thanks for the clarification. Why do we want to match a backslash?



On Sun, Oct 30, 2016 at 10:08 PM +0100, "Stuart Taylor" <notifications@github.com<mailto:notifications@github.com>> wrote:


@Bouncey commented on this pull request.

________________________________

In seed/challenges/01-front-end-development-certification/html5-and-css.json<https://github.com/FreeCodeCamp/FreeCodeCamp/pull/11318>:

> @@ -4584,7 +4584,7 @@
         "assert($(\"h1\").hasClass(\"pink-text\"), 'message: Your <code>h1</code> element should have the class <code>pink-text</code>.');",
         "assert($(\"h1\").hasClass(\"blue-text\"), 'message: Your <code>h1</code> element should have the class <code>blue-text</code>.');",
         "assert($(\"h1\").attr(\"id\") === \"orange-text\", 'message: Your <code>h1</code> element should have the id of <code>orange-text</code>.');",
-        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style[\\s\\S]*?color\\s*?:/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
+        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style\\s*?=\\s*[\\\"]{1}\\s*\\s*\\S*;*\\s*color\\s*?:\\s*?white\\s*;*\\s*[\\\"]{1}/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",


@systimotic<https://github.com/systimotic> these tests are parsed from strings, so \\\" allows for an escaped back-slash, and an escaped double-quote. The quantifier is not needed, see my review ?

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/FreeCodeCamp/FreeCodeCamp/pull/11318>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AG7PV-2pHZxsavhJDs3-wkjR7poTV4-iks5q5Qc1gaJpZM4Kduzm>.

Bouncey commented:
> Why do we want to match a backslash?

The escaped backslash and double quote builds the regEx, so the string `\\\"` when parsed becomes `\"`

dcgit commented:
I originally had capturing groups in there to handle single/double quotes and other scenarios with OR conditions but was asked to remove them. This would have handled situations such as if single quote or double quote was used. Is it okay to add the groupings back in or is there a different method I should be using?

I would like to accommodate the <h1style vs <h1 style as @systimotic pointed out and put my corrections back for single/double quotes and handling spaces with miscellaneous other attributes being added. I will also remove the {1} as I understand it's not needed.

I will update as soon as I can but feel free to correct me if it still doesn't meet expectations, I will also post the test cases I used to validate my reasoning for building the regex as I did.

dcgit committed:
Update html5-and-css.json

Added updated for spaces between h1 and additional attributes, multiple spaces, single or double quotes, removed {1} quanitifier

dcgit commented:
Changes have been completed and the in-browser test I used for the regex validation is included below. I understand this is a basic exercise here but I tried to include what I would think would be common mistakes.
```
var re = new RegExp(/<h1[\s\S]*[\s]+style\s*?=\s*('|")?\s*\s*\S*;*\s*color\s*?:\s*?white\s*;*\s*('|")?/gi);

console.log('<h1 style="color:white">test</h1>'.match(re));

<!--extra spaces-->
console.log('<h1  style = " color : white " >test</h1>'.match(re));
<!--just to be sure, check individual spaces-->
console.log('<h1 style="color:white ">test</h1>'.match(re));
console.log('<h1 style="color: white">test</h1>'.match(re));
console.log('<h1 style="color:white" >test</h1>'.match(re));
console.log('<h1 style="color :white">test</h1>'.match(re));
console.log('<h1 style=" color:white">test</h1>'.match(re));
console.log('<h1 style= "color:white">test</h1>'.match(re));
console.log('<h1 style ="color:white">test</h1>'.match(re));
console.log('<h1  style  =  "  color   :  white  "  >test</h1>'.match(re));
console.log('<h1  style  =  "  color   :  white  ;"  >test</h1>'.match(re));

<!--extra attributes-->
console.log('<h1 class="someclass" style="color:white">test</h1>'.match(re));
console.log('<h1 class="someclass" style="color:white" other stuff="nothing">test</h1>'.match(re));
console.log('<h1 nothing_here style="color:white">test</h1>'.match(re));

<!--try to fake it with duplicate style-->
console.log('<h1 style="color:white;color:green;">test</h1>'.match(re));

<!--do some funky things that the browser would ignore anyway-->
console.log("<h1 style   =  ' color : white   ;;'>test</h1>".match(re));

console.log("The following tests should return null");
<!--spaces that should break things-->
console.log('<h1 style="color:whit e">test</h1>'.match(re));
console.log('<h1 style="color:whit e">test</h1>'.match(re));
console.log('<h1 style="col or:white">test</h1>'.match(re));
console.log('<h1 sty le="color:whit e">test</h1>'.match(re));
console.log('<h 1 style="color:whit e">test</h1>'.match(re));

```

dcgit committed:
made sure color is a separate word

Made sure if anything prefixes color then it is a space or semicolon rather than wordcolor:

Bouncey commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -4584,7 +4584,7 @@
         "assert($(\"h1\").hasClass(\"blue-text\"), 'message: Your <code>h1</code> element should have the class <code>blue-text</code>.');",
         "assert($(\"h1\").attr(\"id\") === \"orange-text\", 'message: Your <code>h1</code> element should have the id of <code>orange-text</code>.');",
-        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style[\\s\\S]*?color\\s*?:/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
+        "assert(code.match(/<h1[\\s\\S]*[\\s]+style\\s*?=\\s*('|\\\")?\\s*([\\s\\S]*;+|)*\\s*color\\s*?:\\s*?white\\s*;*\\s*('|\\\")?/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
```
These tests are parsed from strings, so everything that would required escaping in strings, will need double escaped to make it through the parse. `('|\\\")` should be `(\\'|\\\")`

` ([\\s\\S]*;+|)*` I don't understand this? You are asking it to match everything(zero to unlimited times) before `;`(one to unlimited times) or nothing(zero to unlimited times). `;` is matched by `\S`, so why single it out? 

Your current regExp will match:

```
<h1 kljbvfo;ad


bvo;abv;lkkbnsv;;;;;;;;;;;; style = color: white ;
```

dcgit replied:
The file passed validation so I thought it was ok. I will change to properly escape the single quote.

The intent was to allow any preliminary style attributes as a way of being more flexible such as (style="color:green;font-weight:bold;color:white;") which does render as white text so it should still pass.  Also I did say it does not handle every case of potential format issues such as, it doesn't handle typos of extra < >, mismatched quotes (single vs double), etc. but if the results render on their screen I wouldn't expect to have to check for those. Is this still or would you prefer that I investigate a non-regex solution?

Bouncey commented:
Sorry for the delay in replying to this @dcgit 

Please, keep up the good work :+1:

dcgit committed:
Update html5-and-css.json

BerkeleyTrue commented:
@systimotic @Bouncey Have you been able to review the changes to this PR?

systimotic commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -4584,7 +4584,7 @@
         "assert($(\"h1\").hasClass(\"blue-text\"), 'message: Your <code>h1</code> element should have the class <code>blue-text</code>.');",
         "assert($(\"h1\").attr(\"id\") === \"orange-text\", 'message: Your <code>h1</code> element should have the id of <code>orange-text</code>.');",
-        "assert(code.match(/<h1[\\s\\S]*?style/gi) && code.match(/<h1[\\s\\S]*?style[\\s\\S]*?color\\s*?:/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
+        "assert(code.match(/<h1[\\s\\S]*[\\s]+style\\s*?=\\s*(\\'|\\\")?\\s*([\\s\\S]*;+|)*\\s*color\\s*?:\\s*?white\\s*;*\\s*(\\'|\\\")?/gi), 'message: Give your <code>h1</code> element the inline style of <code>color&#58; white</code>.');",
```
Me again! We're making good progress on this @dcgit, sorry for the delay. There are still two cases of `[\\s\\S]*` in your code. Translated, that says: "match any character that is a space, as well as any character that is not a space, zero or more times". It effectively matches any character. Both of these instances could be removed. 
The `[\\s]+` at the beginning does not need to be wrapped in a character set, `\\s+` should have the same effect.
@Bouncey commented previously on `([\\s\\S]*;+|)*`, which has not changed. Before or after removing the group I talked about above, I can not understand what the goal of this is. I believe it should work fine without this, but I'd really love to learn what this is meant to do.
Looking at the other tests, it seems that double quotes only need a single backslash in front of them to be escaped. Single quotes don't have to be escaped.
`\\s*?` the question mark here and is unnecessary, as the * already means zero or more. There are three cases of this that should be updated.
With the current expression, the quotes are optional. Technically, this would work in real code, but I wouldn't teach anybody this way. Should this stay optional? @Bouncey

raisedadead commented:
Hey @systimotic @dcgit @Bouncey any updates on this?

systimotic commented:
@raisedadead Thanks for picking this up 🙂 
My comments here https://github.com/FreeCodeCamp/FreeCodeCamp/pull/11318#discussion_r88686333 still hold true. I'm still waiting for the changes requested there to be made.
